### PR TITLE
Expose parts of `<cuda/std/memory>`

### PR DESCRIPTION
**Pull request by fbusato**
Fixes https://github.com/NVIDIA/cccl/issues/2499

Provide `std::assume_aligned` backported to previous c++ standards

**TODO:**
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by miscco** [comment]
I have expanded the scope of this PR a bit because we currently do not expose `<memory>` and that has been bothering me a lot in the past.

I exposed those parts of the library we can without issues expose and kept those that are not as easy in a special header

**Comment by fbusato** [comment]
I slightly improved the builtin detection because `__has_builtin` could be problematic. It is not precise in nvc++ (maybe also nvcc) and should not be used with clang < 10.

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 1]
```diff
@@ -87,8 +87,7 @@
 #endif // _CCCL_HAS_BUILTIN(__array_extent)
 
 #if _CCCL_HAS_BUILTIN(__builtin_assume_aligned) || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1923) \
-  || defined(_CCCL_COMPILER_GCC) || defined(_CCCL_COMPILER_NVHPC) || defined(_CCCL_COMPILER_CLANG)              \
-  || defined(_CCCL_CUDA_COMPILER_NVCC)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__cccl/builtin.h]: That is not the right fix, the builtin detection works fine, I did verify that it works will all host compilers including nvc++

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 2]
```diff
@@ -138,7 +137,7 @@
 
 #if _CCCL_CHECK_BUILTIN(builtin_is_constant_evaluated) || (defined(_CCCL_COMPILER_GCC) && _CCCL_GCC_VERSION >= 90000) \
   || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION > 1924 && !defined(_CCCL_CUDACC_BELOW_11_3))                 \
-  || defined(_CCCL_COMPILER_CLANG) || defined(_CCCL_CUDA_COMPILER_NVCC) || defined(_CCCL_COMPILER_NVHPC)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__cccl/builtin.h]: Dito please revert

**Pull request merged**